File: app/src/main/java/is/xyz/mpv/ScalerDialogPreference.java
package is.xyz.mpv;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.content.res.TypedArray;
import android.preference.DialogPreference;
import android.util.AttributeSet;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.Spinner;

import java.util.Arrays;

public class ScalerDialogPreference extends DialogPreference {
    public ScalerDialogPreference(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(context, attrs);
    }
    public ScalerDialogPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }
    public ScalerDialogPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    private View myView;
    private String[] entries;

    private void init(Context context, AttributeSet attrs_) {
        setPersistent(false);
        setDialogLayoutResource(R.layout.scaler_pref);
        TypedArray attrs = context.obtainStyledAttributes(attrs_, R.styleable.ScalerPreferenceDialog);

        // read list of scalers from specified resource
        int res = attrs.getResourceId(R.styleable.ScalerPreferenceDialog_entries, -1);
        entries = context.getResources().getStringArray(res);

        attrs.recycle();
    }

    @Override
    protected void onBindDialogView(View view) {
        super.onBindDialogView(view);
        SharedPreferences prefs = getSharedPreferences();
        myView = view;

        Spinner s = (Spinner) myView.findViewById(R.id.scaler);
        EditText e1 = (EditText) myView.findViewById(R.id.param1);
        EditText e2 = (EditText) myView.findViewById(R.id.param2);

        // populate Spinner and set selected item
        s.setAdapter(new ArrayAdapter<>(getContext(), R.layout.scaler_pref_textview, entries));
        int idx = Arrays.asList(entries).indexOf(prefs.getString(getKey(), ""));
        if(idx != -1)
            s.setSelection(idx, false);

        // populate EditText's
        e1.setText(prefs.getString(getKey() + "_param1", ""));
        e2.setText(prefs.getString(getKey() + "_param2", ""));
    }

    @Override
    protected void onDialogClosed(boolean positiveResult) {
        super.onDialogClosed(positiveResult);
        Editor e = getEditor();

        // save setting values if user presses OK
        if(positiveResult) {
            Spinner s = (Spinner) myView.findViewById(R.id.scaler);
            EditText e1 = (EditText) myView.findViewById(R.id.param1);
            EditText e2 = (EditText) myView.findViewById(R.id.param2);

            e.putString(getKey(), (String) s.getSelectedItem());
            e.putString(getKey() + "_param1", e1.getText().toString());
            e.putString(getKey() + "_param2", e2.getText().toString());
            e.commit();
        }
    }
}

File: app/src/main/java/is/xyz/mpv/config/ScalerDialogPreference.kt
package `is`.xyz.mpv.config

import `is`.xyz.mpv.R
import android.content.Context
import android.preference.DialogPreference
import android.util.AttributeSet
import android.view.View
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.Spinner

class ScalerDialogPreference @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = android.R.attr.dialogPreferenceStyle,
        defStyleRes: Int = 0
): DialogPreference(context, attrs, defStyleAttr, defStyleRes) {
    private var entries: Array<String>

    init {
        isPersistent = false
        dialogLayoutResource = R.layout.scaler_pref

        // read list of scalers from specified resource
        val styledAttrs = context.obtainStyledAttributes(attrs, R.styleable.ScalerPreferenceDialog)
        val res = styledAttrs.getResourceId(R.styleable.ScalerPreferenceDialog_entries, -1)
        entries = context.resources.getStringArray(res)

        styledAttrs.recycle()
    }

    private lateinit var myView: View

    override fun onBindDialogView(view: View) {
        super.onBindDialogView(view)
        myView = view

        val s = myView.findViewById<Spinner>(R.id.scaler)
        val e1 = myView.findViewById<EditText>(R.id.param1)
        val e2 = myView.findViewById<EditText>(R.id.param2)

        // populate Spinner and set selected item
        s.adapter = ArrayAdapter<String>(context, R.layout.scaler_pref_textview, entries)
        val va = sharedPreferences.getString(key, "")
        val idx = entries.indexOf(va)
        if (idx != -1)
            s.setSelection(idx, false)

        // populate EditText's
        e1.setText(sharedPreferences.getString("${key}_param1", ""))
        e2.setText(sharedPreferences.getString("${key}_param2", ""))
    }

    override fun onDialogClosed(positiveResult: Boolean) {
        super.onDialogClosed(positiveResult)

        // save values only if user presses OK
        if (!positiveResult)
            return

        val s = myView.findViewById<Spinner>(R.id.scaler)
        val e1 = myView.findViewById<EditText>(R.id.param1)
        val e2 = myView.findViewById<EditText>(R.id.param2)

        val e = editor // Will create(!) a new SharedPreferences.Editor instance
        e.putString(key, s.selectedItem as String)
        e.putString("${key}_param1", e1.text.toString())
        e.putString("${key}_param2", e2.text.toString())
        e.commit()
    }
}